core/src/main/scala/kafka/controller/ZookeeperClient.scala

```diff
@@ -279,33 +256,132 @@ trait ZNodeChildChangeHandler {
 }
 
 sealed trait AsyncRequest {
-  val path: String
-  val ctx: Any
+  type Response <: AsyncResponse
+  def path: String
+  def ctx: Option[Any]
+  def send(zooKeeper: ZooKeeper)(processResponse: Response => Unit)
```

Comment:
There are a couple reasons for this method:

1. It avoids a bunch of casts in the caller since the method ensures that we return the right type of response and the types are propagated.

2. It clarifies the logic. The previous code may have been clearer to you (you wrote it :)), but I had completely missed how watchers were set in the middle of all the other noise. There was also lots of duplication around updating the mutable state. This approach separates the code for sending the request and creating a response from the code that processes the response.

To address the issues you raise:

1. I agree that it would be better if the `send` method was not visible to API users, I'll see if I can hide it. I didn't think users would be able to use it in unexpected ways since they would not have a `ZooKeeper` instance, but I did mean it to be an implementation detail, so it was an oversight for it to be public.

2. I don't think there's anything wrong with the request knowing how to send itself from a OO or functional perspective.

3. With regards to the `watch` flag, it seemed a bit inconsistent that we had all fields, but that one. Thanks for explaining that the goal is for users not to be able to set it at the request level due to the need to re-register, etc. I'll remove the field.